routing_algorithms/map_matching.hpp

```diff
@@ -0,0 +1,627 @@
+
+        }
+
+        void set_viterbi(const std::vector<std::vector<double>>& viterbi,
+                         const std::vector<std::vector<bool>>& pruned)
+        {
+            // json logger not enabled
+            if (!logger)
+                return;
+
+            for (auto t = 0u; t < viterbi.size(); t++)
+            {
+                for (auto s_prime = 0u; s_prime < viterbi[t].size(); ++s_prime)
+                {
+                    osrm::json::get(*object, "states", t, s_prime, "viterbi") = osrm::json::clamp_float(viterbi[t][s_prime]);
+                    osrm::json::get(*object, "states", t, s_prime, "pruned") = static_cast<unsigned>(pruned[t][s_prime]);
+                }
+            }
+        }
+
+        void add_chosen(const unsigned t, const unsigned s)
+        {
+            // json logger not enabled
+            if (!logger)
+                return;
+
+            osrm::json::get(*object, "states", t, s, "chosen") = true;
+        }
+
+        void add_breakage(const std::vector<bool>& breakage)
+        {
+            // json logger not enabled
+            if (!logger)
+                return;
+
+            osrm::json::get(*object, "breakage") = osrm::json::make_array(breakage);
+        }
+
+        const osrm::json::Logger* logger;
+        osrm::json::Value* object;
+    };
+
+  public:
+    MapMatching(DataFacadeT *facade, SearchEngineData &engine_working_data)
+        : super(facade), engine_working_data(engine_working_data)
+    {
+    }
+
+    void operator()(const Matching::CandidateLists &candidates_list,
+                    const std::vector<FixedPointCoordinate> &trace_coordinates,
+                    const std::vector<unsigned> &trace_timestamps,
+                    const double matching_beta,
+                    const double gps_precision,
+                    Matching::SubMatchingList &sub_matchings) const
+    {
+        BOOST_ASSERT(candidates_list.size() > 0);
+
+        // TODO replace default values with table lookup based on sampling frequency
+        EmissionLogProbability emission_log_probability(gps_precision > 0 ? gps_precision : default_sigma_z);
+        TransitionLogProbability transition_log_probability(matching_beta > 0 ? matching_beta : default_beta);
+
+        HiddenMarkovModel model(candidates_list, emission_log_probability);
+
+        unsigned initial_timestamp = model.initialize(0);
+        if (initial_timestamp == Matching::INVALID_STATE)
+        {
+            return;
+        }
+
+        DebugInfo debug(osrm::json::Logger::get());
+        debug.initialize(candidates_list);
+
+        unsigned breakage_begin = std::numeric_limits<unsigned>::max();
+        std::vector<unsigned> split_points;
+        std::vector<unsigned> prev_unbroken_timestamps;
+        prev_unbroken_timestamps.reserve(candidates_list.size());
+        prev_unbroken_timestamps.push_back(initial_timestamp);
+        for (auto t = initial_timestamp + 1; t < candidates_list.size(); ++t)
```

Comment:
t can't be const, because it needs to be changed by the breakage-recovery-subroutine.